### DISCUSSION TITLE
How to add a new entity properly?

### DISCUSSION THREAD

**Discussion by MilesGibson** [fullstackhero/blazor-starter-kit#347] [Q&A] [answered]
I see that there are a couple of example entities ( Brands/Categories), but zero documentation on how to properly add my own entities into this framework.  This is proving to be a stumbling block for me to be able to use what otherwise looks like a great framework.

TIA,

Miles

**Comment by ajakblackgoat** [comment, marked as answer]
Below are the steps I use when creating a new entity. For details on each step, look in the exiting BlazorHero files.

## Adding new Entity
1. Create and define the new Entity class file in **Domain\Entities** folder .
2. Edit file **Infrastructure\Contexts\BlazorHeroContext.cs** to add ```DbSet<>``` prop and define the entity props in ```OnModelCreating()``` function if required.
3. Edit **Shared\Constants\Permission\Permissions.cs** to add new static class of the new entity containing the permission constants.
4. Edit **Shared\Constants\Application\ApplicationContants.cs** to add the new entity cachekey constant.

## API for the new Entity
1. Create a new FilterSpecification class file in **Application\Specifications** folder.
2. Create a new AddEditCommand class file in **Application\Features** folder.
3. Create a new AddEditCommandValidator class file in **Application\Validators**.
_However, I prefer to add the AddEditCommandValidator class in the same file as the AddEditCommand created above._
4. Create a new DeleteCommand class file in **Application\Features** folder.
5. Create Query class (GetAll, GetAllPaged, GetById, Export, etc as required) and their respective Response class files in **Application\Features** folder. 
_I rarely need the response file because I simply reuse the entity class as the return type for each query. However, if response class is required to minimize the amount of data to be returned, I prefer to define the Response class in the Query files itself instead of seperate file._
6. Create a new controller class file in **Server\Controllers\v1** folder.
7. Create new Manager class and interface files in **Client.Infrastructure\Managers** folder.
8. Create a new Profile class file in **Application\Mappings** folder.
_Optional. Not needed if no mapping is performed in any of the Query files (In case all queries of this specific entity return as entity type instead of response type)._

## UI Files
In **Client\Pages** folder, create all the necessary razor files to manage the entity:
- Page to search/list the entity
- Page to add/edit entity

And finally, edit **Client\Shared\NavMenu.razor** to add a link to access the entity management page.

**Reply by MilesGibson** [reply]
Thanks for this. I trust this will make its way into the documentation.  It is however a lot of steps per entity when you have 50+ entities to build.  So I will have to think on this more.  I currently use DevArt's Entity Developer to generate all of my entities.  I might be able to configure ED to generate the requirements you have set forth...

Thanks again for the excellent documentation.

**Reply by ajakblackgoat** [reply]
Not familiar with DevArt Entity Developer but I believe it is for creating the entity class right?

For my current project, I'm using MySQL 8 as the db. And I design the database table first. Then I use MySql script to generate the entity class codes for the table. At least that takes care of the entity.cs file.

Most of the files are created manually. But not all are required. It really based on the use case of the entity.

Yes, it's troublesome if we have 10's of tables to create their respective entity files. I plan to write a small prog just to generate all those files automatically, but still haven't got the time to do so.

**Comment by MilesGibson** [comment]
Yes that is correct, ED generates both the entity class and the DBContext class (with DBset etc.). 

 

I am going to look at my ED template that I use for generating the entity classes to see if  I can easily generate the other files that you require.  If I can do that, then great.  If not, then I might try what you talked about, build a small program to generate those files into the appropriate folders.  I use Sql Server myself, and of course use the Database First approach as well.

 

I think that if there is a way to generate these entity files programmatically then your framework has a real chance at success.

 

Cheers,

 

Miles

**Comment by MilesGibson** [comment]
Another thought:

 

As I am reviewing the different areas of your application, I realized that it would be a lot easier to generate code in various places by using partial classes.  Using Shared\Constants\Permissions.cs as an example, I would re-structure this to something like this:

 

Shared\Constants\Permissions_Framework.cs  -- where stuff like Users/UserRoles/Permissions/Hangfire etc all go in here, as a ‘public partial static class Permissions’

 

And then there would also be: 

 

Shared\Constants\Permissions_Entities.cs  that would also be defined as ‘public partial static class Permissions’. I could write to this file myself then.

 

Same for ApplicationConstants.cs, and likely most of the other areas as well.

 

Cheers,

 

Miles

**Comment by ajakblackgoat** [comment]
Yes indeed. That's what I'm doing right now for the permission.cs. Changed it to partial and create another partial permission class for business entities.

**Comment by MilesGibson** [comment]
When I look at this:

 

Create Query class (GetAll, GetAllPaged, GetById, Export, etc as required) and their respective Response class files in Application\Features folder.

 

It makes me wonder why you don’t use generics here?

 

  public virtual IEnumerable<T> Get(

                   Expression<Func<T, bool>> filter = null,

                   Func<IQueryable<T>, IOrderedQueryable<T>> orderBy = null,

                   string includeProperties = "")

        {

            try

            {

                // Get the dbSet from the Entity passed in                

                IQueryable<T> query = objectSet;

 

                // Apply the filter

                if (filter != null)

                {

                    query = query.Where(filter);

                }

 

                // Include the specified properties

                foreach (var includeProperty in includeProperties.Split

                    (new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries))

                {

                    query = query.Include(includeProperty);

                }

 

                // Sort

                if (orderBy != null)

                {

                    return orderBy(query).ToList();

                }

                else

                {

                    return query.ToList();

                }

            }

            catch (Exception ex)

            {

                var msg = ex.Message;

                return null;

            }

        }

 

And so on…
